core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

```diff
@@ -326,319 +315,25 @@ public void commit() {
     }
   }
 
-  private void cleanExpiredSnapshots() {
-    // clean up the expired snapshots:
-    // 1. Get a list of the snapshots that were removed
-    // 2. Delete any data files that were deleted by those snapshots and are not in the table
-    // 3. Delete any manifests that are no longer used by current snapshots
-    // 4. Delete the manifest lists
+  ExpireSnapshots withIncrementalCleanup(boolean useIncrementalCleanup) {
+    this.incrementalCleanup = useIncrementalCleanup;
+    return this;
+  }
 
+  private void cleanExpiredSnapshots() {
     TableMetadata current = ops.refresh();
 
-    Set<Long> validIds = Sets.newHashSet();
-    for (Snapshot snapshot : current.snapshots()) {
-      validIds.add(snapshot.snapshotId());
-    }
-
-    Set<Long> expiredIds = Sets.newHashSet();
-    for (Snapshot snapshot : base.snapshots()) {
-      long snapshotId = snapshot.snapshotId();
-      if (!validIds.contains(snapshotId)) {
-        // the snapshot was expired
-        LOG.info("Expired snapshot: {}", snapshot);
-        expiredIds.add(snapshotId);
-      }
-    }
-
-    if (expiredIds.isEmpty()) {
-      // if no snapshots were expired, skip cleanup
-      return;
-    }
-
-    LOG.info("Committed snapshot changes; cleaning up expired manifests and data files.");
-
-    removeExpiredFiles(current.snapshots(), validIds, expiredIds);
-  }
-
-  @SuppressWarnings({"checkstyle:CyclomaticComplexity", "MethodLength"})
-  private void removeExpiredFiles(
-      List<Snapshot> snapshots, Set<Long> validIds, Set<Long> expiredIds) {
-    // Reads and deletes are done using Tasks.foreach(...).suppressFailureWhenFinished to complete
-    // as much of the delete work as possible and avoid orphaned data or manifest files.
-
-    // this is the set of ancestors of the current table state. when removing snapshots, this must
-    // only remove files that were deleted in an ancestor of the current table state to avoid
-    // physically deleting files that were logically deleted in a commit that was rolled back.
-    Set<Long> ancestorIds =
-        Sets.newHashSet(SnapshotUtil.ancestorIds(base.currentSnapshot(), base::snapshot));
-
-    Set<Long> pickedAncestorSnapshotIds = Sets.newHashSet();
-    for (long snapshotId : ancestorIds) {
-      String sourceSnapshotId =
-          base.snapshot(snapshotId).summary().get(SnapshotSummary.SOURCE_SNAPSHOT_ID_PROP);
-      if (sourceSnapshotId != null) {
-        // protect any snapshot that was cherry-picked into the current table state
-        pickedAncestorSnapshotIds.add(Long.parseLong(sourceSnapshotId));
-      }
+    if (incrementalCleanup == null) {
+      incrementalCleanup = current.refs().size() == 1;
```

Comment:
Nice to see that this uses incremental if there's only one ref and incremental isn't specifically set. Good idea.